ykrt/src/compile/jitc_yk/codegen/x64/mod.rs

```diff
@@ -674,38 +736,37 @@ impl<'a> Assemble<'a> {
                     [RegConstraint::InputOutput(lhs), RegConstraint::Input(rhs)],
                 );
                 match size {
-                    1 => dynasm!(self.asm; xor Rb(lhs_reg.code()), Rb(rhs_reg.code())),
-                    2 => dynasm!(self.asm; xor Rw(lhs_reg.code()), Rw(rhs_reg.code())),
-                    4 => dynasm!(self.asm; xor Rd(lhs_reg.code()), Rd(rhs_reg.code())),
-                    8 => dynasm!(self.asm; xor Rq(lhs_reg.code()), Rq(rhs_reg.code())),
+                    1..=8 => dynasm!(self.asm; xor Rq(lhs_reg.code()), Rq(rhs_reg.code())),
                     _ => todo!(),
                 }
             }
             BinOp::UDiv => {
-                let size = lhs.byte_size(self.m);
+                let Ty::Integer(bit_size) = self.m.type_(lhs.tyidx(self.m)) else {
+                    unreachable!()
+                };
                 self.ra
                     .clobber_gp_regs_hack(&mut self.asm, iidx, &[Rq::RDX]);
-                let [_lhs_reg, rhs_reg] = self.ra.assign_gp_regs_avoiding(
+                let [lhs_reg, rhs_reg] = self.ra.assign_gp_regs_avoiding(
                     &mut self.asm,
                     iidx,
                     [
-                        // The quotient is stored in RAX. We don't care about the remainder stored
-                        // in RDX.
+                        // 64-bit unsigned division with idiv operates on RDX:RAX and stores the
+                        // quotient in RAX (we ignore the remainder put into RDX).
                         RegConstraint::InputOutputIntoReg(lhs, Rq::RAX),
                         RegConstraint::Input(rhs),
                     ],
                     RegSet::from(Rq::RDX),
                 );
-                debug_assert_eq!(_lhs_reg, Rq::RAX);
-                // Like SDiv the dividend goes into AX, DX:AX, EDX:EAX, RDX:RAX. But since the
-                // values aren't signed we don't need to sign-extend them and can just zero out
-                // `rdx`.
-                dynasm!(self.asm; xor rdx, rdx);
-                match size {
-                    1 => dynasm!(self.asm; div Rb(rhs_reg.code())),
-                    2 => dynasm!(self.asm; div Rw(rhs_reg.code())),
-                    4 => dynasm!(self.asm; div Rd(rhs_reg.code())),
-                    8 => dynasm!(self.asm; div Rq(rhs_reg.code())),
+                debug_assert_eq!(lhs_reg, Rq::RAX);
+                match bit_size {
+                    1..=64 => {
+                        self.zero_extend_reg64(lhs_reg, u8::try_from(*bit_size).unwrap());
+                        self.zero_extend_reg64(rhs_reg, u8::try_from(*bit_size).unwrap());
+                        dynasm!(self.asm
+                            ; xor rdx, rdx // Zero extend RAX into RDX:RAX.
```

Comment:
This means "put 0 in rdx" I think?

Reply:
Good catch. That's a stale comment. Fixed in f59d229a137f6069551581e65655da620e0f15f2